transaction-view/src/resolved_transaction_view.rs

```diff
@@ -0,0 +1,217 @@
+use {
+    crate::{transaction_data::TransactionData, transaction_view::TransactionView},
+    core::fmt::{Debug, Formatter},
+    solana_sdk::{
+        bpf_loader_upgradeable, ed25519_program,
+        hash::Hash,
+        message::{v0::LoadedAddresses, AccountKeys, TransactionSignatureDetails},
+        pubkey::Pubkey,
+        secp256k1_program,
+    },
+    solana_svm_transaction::{
+        instruction::SVMInstruction, message_address_table_lookup::SVMMessageAddressTableLookup,
+        svm_message::SVMMessage,
+    },
+    std::collections::HashSet,
+};
+
+/// A parsed and sanitized transaction view that has had all address lookups
+/// resolved.
+pub struct ResolvedTransactionView<D: TransactionData> {
+    /// The parsed and sanitized transction view.
+    view: TransactionView<true, D>,
+    /// The resolved address lookups.
+    resolved_addresses: LoadedAddresses,
+    /// A cache for whether an address is writable.
+    writable_cache: Vec<bool>, // TODO: should this be a vec, bitset, or array[256].
+}
+
+impl<D: TransactionData> Debug for ResolvedTransactionView<D> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("ResolvedTransactionView")
+            .field("view", &self.view)
+            .finish()
+    }
+}
+
+impl<D: TransactionData> ResolvedTransactionView<D> {
+    /// Given a parsed and sanitized transaction view, and a set of resolved
+    /// addresses, create a resolved transaction view.
+    pub fn new(
+        view: TransactionView<true, D>,
+        resolved_addresses: LoadedAddresses,
+        reserved_account_keys: &HashSet<Pubkey>, // why does this not use ahash at least?!
+    ) -> Self {
+        let writable_cache =
+            Self::cache_is_writable(&view, &resolved_addresses, reserved_account_keys);
+        Self {
+            view,
+            resolved_addresses,
+            writable_cache,
+        }
+    }
+
+    /// Helper function to check if an address is writable,
+    /// and cache the result.
+    /// This is done so we avoid recomputing the expensive checks each time we call
+    /// `is_writable` - since there is more to it than just checking index.
+    fn cache_is_writable(
+        view: &TransactionView<true, D>,
+        resolved_addresses: &LoadedAddresses,
+        reserved_account_keys: &HashSet<Pubkey>, // why does this not use ahash at least?!
+    ) -> Vec<bool> {
+        // Build account keys so that we can iterate over and check if
+        // an address is writable.
+        let account_keys = AccountKeys::new(view.static_account_keys(), Some(resolved_addresses));
+
+        let mut is_writable_cache = Vec::with_capacity(account_keys.len());
```

Comment:
impl in `SanitizedMessage` was re-calculating all of these per key even though we just loop immediately lol

Reply:
I'll bench different approaches in a follow-up PR to make sure we are the fastest we can be - but for now I moved the `ID` check to inside the existing account key loop so we don't iterate twice.

Reply:
30e635e2caec76d57bbfd6a6f1af2bdc82ea7b24

Reply:
My point was more that it's probably pretty unlikely for a transaction to write-lock a bpf program these days and so normally we rarely do the expensive loader id check. Can we delay the loader check until we know that there is some instruction `ix` where `is_writable_cache[usize::from(ix.program_id_index)] == true`?

Reply:
3122eaa748d3a833bbf598460318e57fc2a62809

Reply:
will create a benchmark as a follow-up. I am somewhat suspicious that the double-iteration and branching may be slower, but happy to be wrong!